librz/util/zip.c

```diff
@@ -107,4 +110,235 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+ * \see rz_deflatew()
+ */
+RZ_API ut8 *rz_deflate(RZ_NONNULL const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
+	rz_return_val_if_fail(src, NULL);
+	return rz_deflatew(src, srcLen, srcConsumed, dstLen, MAX_WBITS + 16);
+}
+
+/**
+ * \brief compress/deflate data to zlib or gzip
+ * \param src source uncompressed bytes
+ * \param srcLen source bytes length
+ * \param srcConsumed consumed source bytes length
+ * \param dstLen compressed bytes length
+ * \param wbits the size of the history buffer (or “window size”), and what header and trailer format is expected.
+ * \return ptr to compressed
+ */
+RZ_API ut8 *rz_deflatew(RZ_NONNULL const ut8 *src, int srcLen, int *srcConsumed, int *dstLen, int wbits) {
+	rz_return_val_if_fail(src, NULL);
+
+	int err = 0;
+	int out_size = 0;
+	ut8 *dst = NULL;
+	ut8 *tmp_ptr;
+	z_stream stream;
+
+	if (srcLen <= 0) {
+		return NULL;
+	}
+
+	memset(&stream, 0, sizeof(z_stream));
+
+	stream.avail_in = srcLen;
+	stream.next_in = (Bytef *)src;
+	stream.zalloc = Z_NULL;
+	stream.zfree = Z_NULL;
+	stream.opaque = Z_NULL;
+
+	if (deflateInit2(&stream, Z_DEFAULT_COMPRESSION, Z_DEFLATED, wbits, 5, Z_DEFAULT_STRATEGY) != Z_OK) {
+		return NULL;
+	}
+
+	do {
+		if (stream.avail_out == 0) {
+			tmp_ptr = realloc(dst, stream.total_out + srcLen);
+			if (!tmp_ptr) {
+				goto err_exit;
+			}
+			dst = tmp_ptr;
+			out_size += srcLen;
+			if (out_size > MAXOUT) {
+				goto err_exit;
+			}
+			stream.next_out = dst + stream.total_out;
+			stream.avail_out = srcLen;
+		}
+		err = deflate(&stream, Z_FINISH);
+		if (err < 0) {
+			RZ_LOG_ERROR("deflate error: %d %s\n", err, gzerr(-err));
+			goto err_exit;
+		}
+	} while (err != Z_STREAM_END);
+
+	if (dstLen) {
+		*dstLen = stream.total_out;
+	}
+	if (srcConsumed) {
+		*srcConsumed = (const ut8 *)stream.next_in - (const ut8 *)src;
+	}
+
+	deflateEnd(&stream);
+	return dst;
+
+err_exit:
+	deflateEnd(&stream);
+	free(dst);
+	return NULL;
+}
+
+/**
+ * \brief deflate uncompressed data in RzBbuffer to zlib or gzipped, use MAX_WBITS as the window size logarithm.
+ * \see rz_deflatew_buf()
+ */
+RZ_API bool rz_deflate_buf(RZ_NONNULL RzBuffer *src, RZ_NONNULL RzBuffer *dst, ut64 block_size, ut8 *src_consumed, ut8 *dst_len) {
+	rz_return_val_if_fail(src && dst, false);
+	return rz_deflatew_buf(src, dst, block_size, src_consumed, dst_len, MAX_WBITS + 16);
+}
+
+/**
+ * \brief deflate data contained in a RzBuffer using zlib
+ * \param src source buffer
+ * \param dst destination buffer
+ * \param block_size block sizes to use while deflating data
+ * \param src_consumed consumed source buffer length
+ * \param dst_len deflated buffer length
+ * \param wbits the size of the history buffer (or “window size”), and what header and trailer format is expected.
+ * \return true if successful; false otherwise
+ */
+RZ_API bool rz_deflatew_buf(RZ_NONNULL RzBuffer *src, RZ_NONNULL RzBuffer *dst, ut64 block_size, ut8 *src_consumed, ut8 *dst_len, int wbits) {
```

Comment:
no need for dst_len, because RzBuffer already has the method rz_buf_size to compute the size.

Reply:
Wouldn't it be consistent to have this parameter (compared to `rz_deflate`)? Just asking.